### PR TITLE
Added ability to use user config file.

### PR DESCRIPTION
**Pull request by omron93**
As discussed in #137 

Added ability to use user config file.
- user is able to create or mount custom config file  in `/etc/mongod.conf` (for example from non-docker MongoDB deployments)
- this config file does not affect MONGODB_DATADIR and MONGODB_REPLICA_NAME

Cleaned up MONGODB env variables and mongod parameters.
- all env var declations were moved into one place
- removed PIDFILE option (mongod stores PID into mongod.lock file in dbpath)
- fixed double declaration of MONGODB_KEYFILE_PATH

### CODE REVIEW

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 1]
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
 
 # setup_keyfile fixes the bug in mounting the Kubernetes 'Secret' volume that
 # mounts the secret files with 'too open' permissions.
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
+  if [ grep -q "keyFile" ${MONGODB_CONFIG_PATH} ]; then
```
Comment (php-coder) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: BTW, you could remove `[` here: `if grep -q "keyFile" ${MONGODB_CONFIG_PATH} ; then` should work the same.

File: 2.6/test/run [hunk 2]
```diff
@@ -280,9 +280,53 @@ function run_change_password_test() {
...
+    local password='password'
+    local admin_password='adminPassword'
+
+    local volume_dir
+    volume_dir=`mktemp -d --tmpdir mongodb-testdata.XXXXX`
+    config_file=$volume_dir/mongod.conf
+    echo """dbpath=/var/lib/mongodb/dbpath
+    unixSocketPrefix = /var/lib/mongodb
+    smallfiles = true
+    nopreallocj = true""" > $config_file
```
Comment (php-coder) [inline 2.6/test/run]: `nopreallocj` <-- is it a typo? (`j` at the end)

Comment (omron93) [inline 2.6/test/run]: Yes. Fixed.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 1 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Thanks.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 3]
```diff
@@ -4,15 +4,23 @@ set -o errexit
...
 export CONTAINER_PORT=27017
 
+# Configuration settings.
+export MONGODB_NOPREALLOC=${MONGODB_NOPREALLOC:-true}
+export MONGODB_SMALLFILES=${MONGODB_SMALLFILES:-true}
+export MONGODB_QUIET=${MONGODB_QUIET:-true}
+
+# Constants used for waiting
+export MAX_ATTEMPTS=60
+export SLEEP_TIME=1
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Why export these variables?

Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: You are right. Exports are needed only for envsub, right?

So removed other unnecessary exports.

Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Yeah, MAX_ATTEMPTS and SLEEP_TIME are globals only used in this file.

Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Also MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, or not?

/usr/share/container-scripts/mongodb/common.sh is sourced in every run or execed file and config file does not contain them... so it is necessary to export them.

Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: > Also MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, or not?

Here's where Bash haunts our souls again... :cry: 
I'm not sure if it is necessary to export in that case, I guess not necessary, need to try it out.

Here are the occurrences of MONGODB_CONFIG_PATH:

``` console
$ git grep -n MONGODB_CONFIG_PATH upstream/master -- 2.4
upstream/master:2.4/root/usr/bin/run-mongod:67:envsubst < ${CONTAINER_SCRIPTS_PATH}/mongodb.conf.template > $MONGODB_CONFIG_PATH
upstream/master:2.4/root/usr/bin/run-mongod:71:mongo_common_args="-f $MONGODB_CONFIG_PATH --oplogSize 64"
upstream/master:2.4/root/usr/share/container-scripts/mongodb/common.sh:11:export MONGODB_CONFIG_PATH=/etc/mongod.conf
upstream/master:2.4/root/usr/share/container-scripts/mongodb/initiate_replica.sh:72:mongod -f ${MONGODB_CONFIG_PATH} --shutdown &>/dev/null
```

Trying it out:

``` console
$ cat foo
#!/bin/bash

set -o errexit
set -o nounset
set -o pipefail

source bar

echo $V_EXPORTED
echo $V_NOT_EXPORTED
$ cat bar
#!/bin/bash

export V_EXPORTED="exported var"
V_NOT_EXPORTED="not exported var"
$ ./foo 
exported var
not exported var
```

Conclusion: no need to export MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, @omron93 agree?

BTW, MONGODB_KEYFILE_PATH is defined/exported both in `2.4/root/usr/bin/run-mongod:21` and `2.4/root/usr/share/container-scripts/mongodb/common.sh:13`, we should likely get rid of one of those lines. Notice that $HOME == /var/lib/mongodb.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 4]
```diff
@@ -8,8 +8,8 @@ set -o pipefail
 # because mongodb.conf lives in /var/lib/mongodb/ and we don't want a volume to
 # override it.
 export MONGODB_DATADIR=/var/lib/mongodb/data
-export MONGODB_CONFIG_PATH=/etc/mongod.conf
-export MONGODB_KEYFILE_PATH="${HOME}/keyfile"
+MONGODB_CONFIG_PATH=/etc/mongod.conf
+MONGODB_KEYFILE_PATH="${HOME}/keyfile"
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: How about organizing it such that exported and non-exported variables are separated?
I think it would be less puzzling for readers.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 5]
```diff
@@ -18,8 +18,8 @@ export MONGODB_SMALLFILES=${MONGODB_SMALLFILES:-true}
 export MONGODB_QUIET=${MONGODB_QUIET:-true}
 
 # Constants used for waiting
-export MAX_ATTEMPTS=60
-export SLEEP_TIME=1
+MAX_ATTEMPTS=60
+SLEEP_TIME=1
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: We may want to make these `readonly`, if they're really constants not intended to be tweaked.
Or if we want to make it possible to customize them, maybe something like `MAX_ATTEMPTS={MAX_ATTEMPTS:-60}`, WDYT?

Since we never tweaked these values (not possible with current implementation), it'd vote for `readonly`.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 3 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: > Conclusion: no need to export MONGODB_CONFIG_PATH and MONGODB_KEYFILE_PATH, 

Agree.

> BTW, MONGODB_KEYFILE_PATH is defined/exported both in 2.4/root/usr/bin/run-mongod:21 and 2.4/root/usr/share/container-scripts/mongodb/common.sh:13, we should likely get rid of one of those lines. Notice that $HOME == /var/lib/mongodb.

This is not true in this PR.

File: 2.4/root/usr/bin/run-mongod [hunk 6]
```diff
@@ -6,20 +6,6 @@ set -o pipefail
...
-# override it.
-export MONGODB_DATADIR=/var/lib/mongodb/data
-
-# Configuration settings.
-export MONGODB_NOPREALLOC=${MONGODB_NOPREALLOC:-true}
-export MONGODB_SMALLFILES=${MONGODB_SMALLFILES:-true}
-export MONGODB_QUIET=${MONGODB_QUIET:-true}
-
-
-export MONGODB_KEYFILE_SOURCE_PATH="/var/run/secrets/mongo/keyfile"
```
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod:20]: I can't find where else this is defined?

File: 2.4/root/usr/bin/run-mongod [hunk 7]
```diff
@@ -6,20 +6,6 @@ set -o pipefail
...
-export MONGODB_KEYFILE_PATH="${HOME}/keyfile"
```
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod:21]: I see this was defined both in `common.sh` and `run-mongod`, removal looks good.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 4 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: This sounds good.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 5 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Agree. Added readonly.

File: 2.4/root/usr/bin/run-mongod (hunk 6 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod:20]: Nowhere. I can't fine where it is used?

So I removed it.

Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod:20]: Seems right to remove it. It was originally added together with the replication example in https://github.com/openshift/mongodb/pull/17, but later its usage was removed in https://github.com/openshift/mongodb/pull/45 and the export was left over. Thanks for spotting it!

**Comment by omron93** [comment]
@rhcarvalho @php-coder Any other suggestions?

File: 2.4/root/usr/bin/run-mongod [hunk 8]
```diff
@@ -46,8 +32,8 @@ function cleanup() {
     mongo_remove
   fi
   echo "=> Shutting down MongoDB server ..."
-  if [ -f "${MONGODB_PID_FILE}" ]; then
-    kill -2 $(cat ${MONGODB_PID_FILE})
+  if [[ -s ${MONGODB_DATADIR}/mongod.lock ]]; then
+    kill -2 $(cat ${MONGODB_DATADIR}/mongod.lock)
```
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: How is this change related to user ability to provide a config file?

File: 2.4/root/usr/bin/run-mongod [hunk 9]
```diff
@@ -63,12 +49,15 @@ if [ $# -gt 0 ] && [ "$1" == "initiate" ]; then
...
+# If user provides own config file use it and do not generate new one
+if [ ! -s $MONGODB_CONFIG_PATH ]; then
+  # Generate config file for MongoDB
+  envsubst < ${CONTAINER_SCRIPTS_PATH}/mongodb.conf.template > $MONGODB_CONFIG_PATH
+fi
 
 # Need to cache the container address for the cleanup
 cache_container_addr
-mongo_common_args="-f $MONGODB_CONFIG_PATH --oplogSize 64"
+mongo_common_args="-f $MONGODB_CONFIG_PATH --dbpath $MONGODB_DATADIR"
```
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: Why now we need to pass a custom dbpath?

File: 2.4/test/run [hunk 10]
```diff
@@ -280,9 +280,53 @@ function run_change_password_test() {
...
+    echo "  Testing config file mount"
+    local database='db'
+    local user='user'
+    local password='password'
+    local admin_password='adminPassword'
+
+    local volume_dir
+    volume_dir=`mktemp -d --tmpdir mongodb-testdata.XXXXX`
+    config_file=$volume_dir/mongod.conf
+    echo """dbpath=/var/lib/mongodb/dbpath
```
Comment (rhcarvalho) [inline 2.4/test/run]: Why the triple double quotes here?!

Looks like Python, but this is not Python :P

File: 2.4/root/usr/bin/run-mongod [hunk 11]
```diff
@@ -114,8 +103,7 @@ else
   # `cleanup` references it.
   (
     unset_env_vars
-    mongod $mongo_common_args --replSet ${MONGODB_REPLICA_NAME} \
-      --keyFile ${MONGODB_KEYFILE_PATH} ${auth_args}
+    mongod $mongo_common_args --replSet ${MONGODB_REPLICA_NAME} ${auth_args}
```
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: I see this was already lacking proper quoting... but maybe it's a good time to add it?

`MONGODB_REPLICA_NAME` should surely be double-quoted, while `mongo_common_args` and `auth_args` we may want them to expand to multiple things... seems weird to have 2 vars that expand to arbitrary things.
Even then, maybe we can quote and treat them as arrays with the magical @ syntax...

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 12]
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
...
 # mounts the secret files with 'too open' permissions.
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
+  if grep -q "keyFile" ${MONGODB_CONFIG_PATH}; then
+    exit 0
+  fi
   if [ -z "${MONGODB_KEYFILE_VALUE-}" ]; then
-    echo "ERROR: You have to provide the 'keyfile' value in $MONGODB_KEYFILE_VALUE"
+    echo "ERROR: You have to provide the 'keyfile' value in MONGODB_KEYFILE_VALUE"
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh:270]: Don't know why this is still showing up here, but it is already in master... could you please rebase your PR (and probably squash the commits)?

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 13]
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
 
 # setup_keyfile fixes the bug in mounting the Kubernetes 'Secret' volume that
 # mounts the secret files with 'too open' permissions.
+# add --keyFile argument to mongo_common_args
 function setup_keyfile() {
+  # If user specify keyFile in cofig file do not use generated keyFile
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: s/cofig/config

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh [hunk 14]
```diff
@@ -253,11 +261,17 @@ function mongo_reset_passwords() {
...
+  if grep -q "keyFile" ${MONGODB_CONFIG_PATH}; then
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: this is a very broad grep... I'd prefer something more specific to avoid matching the string "keyFile" in some arbitrary place of the config, like a value or comment, not a key.

File: 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template [hunk 15]
```diff
@@ -1,7 +1,6 @@
 # mongodb.conf
 
-port = 27017
-pidfilepath = ${MONGODB_PID_FILE}
```
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template:4]: Why are we removing MONGODB_PID_FILE? If we are to remove it, we should update the README / docs too (there might be OpenShift docs referring to it).

File: 2.4/test/run (hunk 10 above)
Comment (omron93) [inline 2.4/test/run]: You are right. Fixed.

File: 2.4/root/usr/bin/run-mongod (hunk 8 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: If the image provides ability to mount config file scripts should avoid getting values from it. (I thought that OpenShift suggested not to do this to avoid bugs for some user weird config files)

MONGODB_PID_FILE is one example of it. The same value is stored in file `mongod.lock` in dbpath. So no reason to get value of MONGODB_PID_FILE from config file.

File: 2.4/root/usr/bin/run-mongod (hunk 9 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: To avoid getting values from config file - https://github.com/openshift/mongodb/pull/147#discussion_r61854214

`dbpath` is one example and it is really necessary for MongoDB scripts. Also for better performance database should be stored in the docker volume. And user mounting config files from "classical" deployments don't have to be aware of it.

So explicit `--dbpath` overrides value from config file.

File: 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template (hunk 15 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template:4]: Same as in https://github.com/openshift/mongodb/pull/147#discussion_r61854214 MongoDB scripts won't use this variable. 

@rhcarvalho Who can use this? Is it somehow possible use this file with PID in other dependent applications in different containers?

What docs please? I can't find any...

File: 2.4/root/usr/bin/run-mongod (hunk 11 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: I agree with quoting  `MONGODB_REPLICA_NAME`. I don't know why `auth_args` are separated (Also this is not necessary for MongoDB 2.6+ because `--keyFile` implies auth). So `--auth` could be simply added to `mongo_common_args` for MongoDB 2.4.

@rhcarvalho Are you fine with doing it in this PR?

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 14 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Now matching only lines starting with `keyFile`. Should be valid for both MongoDB config file formats.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 13 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh]: Fixed.

File: 2.4/root/usr/share/container-scripts/mongodb/common.sh (hunk 12 above)
Comment (omron93) [inline 2.4/root/usr/share/container-scripts/mongodb/common.sh:270]: Rebased.

File: 2.4/root/usr/bin/run-mongod (hunk 9 above)
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: Hmm, wasn't it the whole point to allow users to pass in whatever config they want? Weird to me to overwrite the dbpath.

Our default config should point to the correct dir, and perhaps we document this concern that you mention for users providing their own custom config file?

File: 2.4/root/usr/bin/run-mongod (hunk 8 above)
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: Maybe there should be some thought about backwards-compat before moving on with this change?

Is MONGODB_PID_FILE being read from the config file?!

Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: What "backward-compat" do you mean?

`MONGODB_PID_FILE` before this PR is specified in `/usr/share/container-scripts/mongodb/common.sh` and it is used to stop container. In this PR `MONGODB_PID_FILE` is not used in scripts, so no need have it present in scripts.
However if you know some usecase when MONGODB_PID_FILE should be created with default config file. It could be easily added back to config.tepmlate.

File: 2.4/root/usr/bin/run-mongod (hunk 9 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: Yes, it was. And I agree with you.

I would be happier if our scripts would not rely on this. I will try to remove this dependency.

File: 2.4/root/usr/bin/run-mongod (hunk 8 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: @rhcarvalho Do you know why `cleanup()` uses `kill` or `pkill` instead of `mongod --shutdown`?

File: 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template (hunk 15 above)
Comment (rhcarvalho) [inline 2.4/root/usr/share/container-scripts/mongodb/mongodb.conf.template:4]: If it's just used internally then I see no problems with removing it.

File: 2.4/root/usr/bin/run-mongod (hunk 8 above)
Comment (rhcarvalho) [inline 2.4/root/usr/bin/run-mongod]: > Do you know why cleanup() uses kill or pkill instead of mongod --shutdown?

I don't know, probably using `--shutdown` would be preferred. But again, it's easier to review PRs that do one thing.
That change would be much easier to review in a PR of itself IMHO.

Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: > it's easier to review PRs that do one thing.
> That change would be much easier to review in a PR of itself IMHO.

Yes. But it depends on graininess. If "every line change" is submited in separate PR, it really makes improvements harder and makes thinks slower.

However, there is a new PR #155

File: 2.4/root/usr/bin/run-mongod (hunk 11 above)
Comment (omron93) [inline 2.4/root/usr/bin/run-mongod]: New PR for this #156

File: 2.4/root/usr/bin/run-mongod (hunk 9 above)
Comment (omron93) [reply 2.4/root/usr/bin/run-mongod]: Necessity to override user dbpath was removed in #157. So removed from this PR.

**Comment by omron93** [comment]
Rebased and also added 3.2 version.

So now:

>    Added ability to use user config file.
> - user is able to create or mount custom config file in /etc/mongod.conf (for example from non-docker MongoDB deployments)
> - this config file does not affect MONGODB_REPLICA_NAME (possibility to use user config file is not aimed at replicated environment)

@hhorak @bparees @rhcarvalho If there are no other suggestions please merge.

**Comment by bparees** [comment]
lgtm, but please update the READMEs to describe the new supported behavior for supplying a config file:
https://github.com/sclorg/mongodb-container/blob/master/2.4/README.md
https://github.com/sclorg/mongodb-container/blob/master/2.6/README.md
https://github.com/sclorg/mongodb-container/blob/master/3.2/README.md

**Comment by bparees** [comment]
[test]

**Comment by omron93** [comment]
@bparees README updated.

Test failed, but it seems that it is not a problem in this PR.

**Comment by bparees** [comment]
i'm not sure the failure isn't the fault of this PR:

```
+ docker run --rm openshift/mongodb-24-rhel7-candidate mongo db --host 172.17.0.2 -u user -p password --eval 'db.getSiblingDB('\''test_database'\'');'
exception: connect failed
+ status=1
+ set -e
+ '[' 1 -eq 0 ']'
+ sleep 2
+ echo '  Giving up: Failed to connect. Logs:'
```

seems like it's not able to connect.  are you able to run `make test` on this PR locally successfully?

**Comment by omron93** [comment]
It is fix in this PR, I am sorry.

Should be fixed. Please retest.

**Comment by bparees** [comment]
[test]

File: 2.4/README.md [hunk 16]
```diff
@@ -58,6 +58,14 @@ will be started. If you are re-attaching the volume to another container, the
 creation of the database user and admin user will be skipped and only the
 MongoDB daemon will be started.
 
+Custom configuration file
+---------------------------------
+
+It is allowed to use custom configuration file for mongod server.
+
+To use custom configuration file in container it has to be mounted into `/etc/mongod.conf`. For example to use configuration file stored in `/home/user` directory use this option for `docker run` command: `-v /home/user/mongod.conf:/etc/mongod.conf:Z`.
```
Comment (bparees) [inline 2.4/README.md:66]: "Providing a custom configuration file supercedes the individual configuration environment variable values."

**Comment by bparees** [comment]
one readme nit and lgtm, @hhorak ?

File: 2.4/README.md (hunk 16 above)
Comment (omron93) [reply 2.4/README.md:66]: Look good, thanks. Added.

**Comment by hhorak** [comment]
@omron93, after 30-upg was merged, could you do the same changes for that version as well, pls?
Otherwise looks good to me, so LGTM.

**Comment by omron93** [comment]
Rebased and added 30-upg.

@bparees @hhorak Please merge.
